pkg/configurator/crd_client.go

```diff
@@ -0,0 +1,183 @@
+package configurator
+
+import (
+	"strings"
+
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/openservicemesh/osm/pkg/announcements"
+	"github.com/openservicemesh/osm/pkg/apis/config/v1alpha1"
+	"github.com/openservicemesh/osm/pkg/gen/client/config/clientset/versioned"
+	informers "github.com/openservicemesh/osm/pkg/gen/client/config/informers/externalversions"
+	k8s "github.com/openservicemesh/osm/pkg/kubernetes"
+	"github.com/openservicemesh/osm/pkg/kubernetes/events"
+)
+
+const (
+	meshConfigInformerName = "MeshConfig"
+	meshConfigProviderName = "OSM"
+)
+
+func newConfiguratorWithCRDClient(meshConfigClientSet versioned.Interface, stop <-chan struct{}, osmNamespace string) *CRDClient {
+	informerFactory := informers.NewSharedInformerFactoryWithOptions(
+		meshConfigClientSet,
+		k8s.DefaultKubeEventResyncInterval,
+		informers.WithNamespace(osmNamespace),
+	)
+	informer := informerFactory.Config().V1alpha1().MeshConfigs().Informer()
+	crdClient := CRDClient{
+		informer:     informer,
+		cache:        informer.GetStore(),
+		cacheSynced:  make(chan interface{}),
+		osmNamespace: osmNamespace,
+	}
+
+	// configure listener
+	eventTypes := k8s.EventTypes{
+		Add:    announcements.MeshConfigAdded,
+		Update: announcements.MeshConfigUpdated,
+		Delete: announcements.MeshConfigDeleted,
+	}
+	informer.AddEventHandler(k8s.GetKubernetesEventHandlers(meshConfigInformerName, meshConfigProviderName, nil, eventTypes))
+
+	// start listener
+	crdClient.startMeshConfigListener(stop)
+
+	// start informer
+	go informer.Run(stop)
+
+	return &crdClient
+}
+
+// Listens to ConfigMap events and notifies dispatcher to issue config updates to the envoys based
+// on config seen on the configmap
+func (c *CRDClient) startMeshConfigListener(stop <-chan struct{}) {
+	// Create the subscription channel synchronously
+	cfgSubChannel := events.GetPubSubInstance().Subscribe(
+		announcements.MeshConfigAdded,
+		announcements.MeshConfigDeleted,
+		announcements.MeshConfigUpdated,
+	)
+
+	go func() {
+		// Defer unsubscription on async routine exit
+		defer events.GetPubSubInstance().Unsub(cfgSubChannel)
+
+		for {
+			select {
+			case msg := <-cfgSubChannel:
+				psubMsg, ok := msg.(events.PubSubMessage)
+				if !ok {
+					log.Error().Msgf("Could not cast pubsub message")
```

Comment:
This is technically type assertion and not casting. (I know this subtle error is in a few other places as well - it'd be nice to correct these too).
```suggestion
					log.Error().Msgf("Type assertion failed for PubSubMessage")
```